File: assets/src/shared/lib/crypto/suite.ts
import { blake3Base64Url } from "./hash";
import { canonicalizeStrictBytes, compareUtf8 } from "./jcs";

export const SUITE_IDS = {
  SIGNED_PQ_HYBRID_WRAP:
    "refmd-v2-draft-ietf-hpke-pq-04-mlkem768-x25519-hkdfsha256-chacha20poly1305-ed25519-mldsa65",
  HYBRID_SIGNATURE: "refmd-v2-hybrid-signature-ed25519-mldsa65",
  INITIAL_AKE: "refmd-v2-initial-ake-mlkem768-x25519-hkdfsha256-ed25519-mldsa65",
  INITIAL_DELIVERY: "refmd-v2-initial-delivery-xchacha20poly1305",
} as const;

export const CURRENT_PROTOCOL_VERSION = 1;
export const CURRENT_SUITE_RANK = 1000;
export const CURRENT_SUITE_POLICY_VERSION = 1;

const CURRENT_ALLOWED_SUITE_IDS = Object.values(SUITE_IDS).sort(compareUtf8);
const CURRENT_REQUIRED_COMPONENTS = ["ed25519", "mldsa65", "mlkem768", "x25519"].sort(compareUtf8);

export interface SuitePolicy {
  suite_policy_version: number;
  min_suite_rank: number;
  allowed_suite_ids: string[];
  required_components: string[];
  allowed_suite_ids_hash?: string;
}

export interface PinnedSuitePolicy {
  suite_policy_version: number;
  min_suite_rank: number;
  allowed_suite_ids_hash: string;
}

export function currentSuitePolicy(): SuitePolicy {
  return {
    suite_policy_version: CURRENT_SUITE_POLICY_VERSION,
    min_suite_rank: CURRENT_SUITE_RANK,
    allowed_suite_ids: [...CURRENT_ALLOWED_SUITE_IDS],
    required_components: [...CURRENT_REQUIRED_COMPONENTS],
    allowed_suite_ids_hash: canonicalAllowedSuiteIdsHash({
      allowed_suite_ids: [...CURRENT_ALLOWED_SUITE_IDS],
    }),
  };
}

export function assertProtocolVersion(protocolVersion: number): void {
  if (protocolVersion !== CURRENT_PROTOCOL_VERSION) {
    throw new Error("protocol_version_not_allowed");
  }
}

export function assertKnownSuiteId(
  suiteId: string,
  policy: SuitePolicy = currentSuitePolicy(),
): void {
  assertSuitePolicyShape(policy);
  if (!policy.allowed_suite_ids.includes(suiteId)) {
    throw new Error("suite_id_not_allowed");
  }
}

export function assertSuiteRankAllowed(
  suiteId: string,
  suiteRank: number,
  policy: SuitePolicy = currentSuitePolicy(),
): void {
  assertKnownSuiteId(suiteId, policy);
  if (suiteRank < policy.min_suite_rank || suiteRank !== CURRENT_SUITE_RANK) {
    throw new Error("suite_rank_not_allowed");
  }
}

export function assertRequiredComponents(policy: SuitePolicy): void {
  assertCanonicalSortedUnique(policy.required_components, "required_components");
  if (
    policy.required_components.length !== CURRENT_REQUIRED_COMPONENTS.length ||
    policy.required_components.some((component, i) => component !== CURRENT_REQUIRED_COMPONENTS[i])
  ) {
    throw new Error("required_components_mismatch");
  }
}

export function canonicalAllowedSuiteIdsHash(
  policy: Pick<SuitePolicy, "allowed_suite_ids">,
): string {
  assertCanonicalSortedUnique(policy.allowed_suite_ids, "allowed_suite_ids");
  return blake3Base64Url(canonicalizeStrictBytes({ allowed_suite_ids: policy.allowed_suite_ids }));
}

export function assertPinnedSuitePolicy(policy: SuitePolicy, pinned: PinnedSuitePolicy): void {
  assertSuitePolicyShape(policy);
  if (policy.suite_policy_version < pinned.suite_policy_version) {
    throw new Error("suite_policy_version_rollback");
  }
  if (policy.min_suite_rank < pinned.min_suite_rank) {
    throw new Error("min_suite_rank_rollback");
  }
  if (canonicalAllowedSuiteIdsHash(policy) !== pinned.allowed_suite_ids_hash) {
    throw new Error("allowed_suite_ids_hash_mismatch");
  }
}

function assertSuitePolicyShape(policy: SuitePolicy): void {
  assertCanonicalSortedUnique(policy.allowed_suite_ids, "allowed_suite_ids");
  const hash = canonicalAllowedSuiteIdsHash(policy);
  if (policy.allowed_suite_ids_hash !== undefined && policy.allowed_suite_ids_hash !== hash) {
    throw new Error("allowed_suite_ids_hash_mismatch");
  }
  if (!CURRENT_ALLOWED_SUITE_IDS.every((suiteId) => policy.allowed_suite_ids.includes(suiteId))) {
    throw new Error("allowed_suite_ids_incomplete");
  }
  if (policy.allowed_suite_ids.length !== CURRENT_ALLOWED_SUITE_IDS.length) {
    throw new Error("allowed_suite_ids_unknown");
  }
  assertRequiredComponents(policy);
}

function assertCanonicalSortedUnique(values: string[], field: string): void {
  const seen = new Set<string>();
  for (let i = 0; i < values.length; i += 1) {
    const value = values[i]!;
    if (seen.has(value)) throw new Error(`${field}_duplicate`);
    seen.add(value);
    if (i > 0 && compareUtf8(values[i - 1]!, value) >= 0) {
      throw new Error(`${field}_not_canonical`);
    }
  }
}

File: lib/refmd/crypto/suite.ex
defmodule RefMD.Crypto.Suite do
  @moduledoc false

  alias RefMD.Crypto.Hash
  alias RefMD.Crypto.JCS

  @signed_pq_hybrid_wrap "refmd-v2-draft-ietf-hpke-pq-04-mlkem768-x25519-hkdfsha256-chacha20poly1305-ed25519-mldsa65"
  @hybrid_signature "refmd-v2-hybrid-signature-ed25519-mldsa65"
  @initial_ake "refmd-v2-initial-ake-mlkem768-x25519-hkdfsha256-ed25519-mldsa65"
  @initial_delivery "refmd-v2-initial-delivery-xchacha20poly1305"

  @protocol_version 1
  @suite_rank 1000
  @suite_policy_version 1
  @allowed_suite_ids Enum.sort([
                       @signed_pq_hybrid_wrap,
                       @hybrid_signature,
                       @initial_ake,
                       @initial_delivery
                     ])
  @required_components Enum.sort(["ed25519", "mldsa65", "mlkem768", "x25519"])

  def current_suite_policy do
    policy = %{
      "suite_policy_version" => @suite_policy_version,
      "min_suite_rank" => @suite_rank,
      "allowed_suite_ids" => @allowed_suite_ids,
      "required_components" => @required_components
    }

    Map.put(policy, "allowed_suite_ids_hash", canonical_allowed_suite_ids_hash(policy))
  end

  def assert_protocol_version!(@protocol_version), do: :ok
  def assert_protocol_version!(_), do: raise(ArgumentError, "protocol_version_not_allowed")

  def initial_ake_suite_id, do: @initial_ake

  def initial_delivery_suite_id, do: @initial_delivery

  def current_suite_rank, do: @suite_rank

  def assert_known_suite_id!(suite_id, policy \\ current_suite_policy()) do
    assert_suite_policy_shape!(policy)

    if suite_id in policy["allowed_suite_ids"] do
      :ok
    else
      raise ArgumentError, "suite_id_not_allowed"
    end
  end

  def assert_suite_rank_allowed!(suite_id, suite_rank, policy \\ current_suite_policy()) do
    assert_known_suite_id!(suite_id, policy)

    if suite_rank < policy["min_suite_rank"] or suite_rank != @suite_rank do
      raise ArgumentError, "suite_rank_not_allowed"
    end

    :ok
  end

  def assert_required_components!(%{"required_components" => components}) do
    assert_canonical_sorted_unique!(components, "required_components")

    if components == @required_components do
      :ok
    else
      raise ArgumentError, "required_components_mismatch"
    end
  end

  def canonical_allowed_suite_ids_hash(%{"allowed_suite_ids" => allowed_suite_ids}) do
    assert_canonical_sorted_unique!(allowed_suite_ids, "allowed_suite_ids")
    Hash.blake3_base64url(JCS.canonical_bytes!(%{"allowed_suite_ids" => allowed_suite_ids}))
  end

  def assert_pinned_suite_policy!(policy, pinned_policy) do
    assert_suite_policy_shape!(policy)

    cond do
      policy["suite_policy_version"] < pinned_policy["suite_policy_version"] ->
        raise ArgumentError, "suite_policy_version_rollback"

      policy["min_suite_rank"] < pinned_policy["min_suite_rank"] ->
        raise ArgumentError, "min_suite_rank_rollback"

      canonical_allowed_suite_ids_hash(policy) != pinned_policy["allowed_suite_ids_hash"] ->
        raise ArgumentError, "allowed_suite_ids_hash_mismatch"

      true ->
        :ok
    end
  end

  defp assert_suite_policy_shape!(policy) do
    assert_canonical_sorted_unique!(policy["allowed_suite_ids"], "allowed_suite_ids")

    hash = canonical_allowed_suite_ids_hash(policy)

    if Map.has_key?(policy, "allowed_suite_ids_hash") && policy["allowed_suite_ids_hash"] != hash do
      raise ArgumentError, "allowed_suite_ids_hash_mismatch"
    end

    if not Enum.all?(@allowed_suite_ids, &(&1 in policy["allowed_suite_ids"])) do
      raise ArgumentError, "allowed_suite_ids_incomplete"
    end

    if length(policy["allowed_suite_ids"]) != length(@allowed_suite_ids) do
      raise ArgumentError, "allowed_suite_ids_unknown"
    end

    assert_required_components!(policy)
  end

  defp assert_canonical_sorted_unique!(values, field) when is_list(values) do
    if Enum.uniq(values) != values do
      raise ArgumentError, "#{field}_duplicate"
    end

    if Enum.sort(values) != values do
      raise ArgumentError, "#{field}_not_canonical"
    end

    :ok
  end
end
